examples/src/main/scala/org/apache/spark/examples/mllib/PowerIterationClusteringExample.scala

```diff
@@ -0,0 +1,149 @@
+ * with each of the input circles.
+ *
+ * Run with
+ * {{{
+ * ./bin/run-example mllib.PowerIterationClusteringExample [options]
+ *
+ * Where options include:
+ *   k:  Number of circles/ clusters
+ *   n:  Number of sampled points on innermost circle.. There are proportionally more points
+ *      within the outer/larger circles
+ *   numIterations:   Number of Power Iterations
+ *   outerRadius:  radius of the outermost of the concentric circles
+ * }}}
+ *
+ * Here is a sample run and output:
+ *
+ * ./bin/run-example mllib.PowerIterationClusteringExample
+ * -k 3 --n 30 --numIterations 15
+ *
+ * Cluster assignments: 1 -> [0,1,2,3,4],2 -> [5,6,7,8,9,10,11,12,13,14],
+ * 0 -> [15,16,17,18,19,20,21,22,23,24,25,26,27,28,29]
+ *
+ *
+ * If you use it as a template to create your own app, please use `spark-submit` to submit your app.
+ */
+object PowerIterationClusteringExample {
+
+  def main(args: Array[String]) {
+    val defaultParams = Params()
+
+    val parser = new OptionParser[Params]("PIC Circles") {
+      head("PowerIterationClusteringExample: an example PIC app using concentric circles.")
+      opt[Int]('k', "k")
+          .text(s"number of circles (/clusters), default: ${defaultParams.k}")
+          .action((x, c) => c.copy(k = x))
+      opt[Int]('n', "n")
+          .text(s"number of points, default: ${defaultParams.numPoints}")
+          .action((x, c) => c.copy(numPoints = x))
+      opt[Int]("numIterations")
+          .text(s"number of iterations, default: ${defaultParams.numIterations}")
+          .action((x, c) => c.copy(numIterations = x))
+    }
+
+    parser.parse(args, defaultParams).map { params =>
+      run(params)
+    }.getOrElse {
+      sys.exit(1)
+    }
+  }
+
+  def run(params: Params) {
+    val conf = new SparkConf()
+        .setMaster("local")
+        .setAppName(s"PowerIterationClustering with $params")
+    val sc = new SparkContext(conf)
+
+    Logger.getRootLogger.setLevel(Level.WARN)
+
+    val circlesRdd = generateCirclesRdd(sc, params.k, params.numPoints, params.outerRadius)
+    val model = new PowerIterationClustering()
+        .setK(params.k)
+        .setMaxIterations(params.numIterations)
+        .run(circlesRdd)
+
+    val clusters = model.assignments.collect.groupBy(_._2).mapValues(_.map(_._1))
+    val assignments = clusters.toList.sortBy { case (k, v) => v.length}
+    val assignmentsStr = assignments
+        .map { case (k, v) => s"$k -> ${v.sorted.mkString("[", ",", "]")}"}.mkString(",")
+    println(s"Cluster assignments: $assignmentsStr")
+
+    sc.stop()
+  }
+
+  def generateCirclesRdd(sc: SparkContext,
+      nCircles: Int = 3,
+      nPoints: Int = 30,
+      outerRadius: Double): RDD[(Long, Long, Double)] = {
+
+    val radii = for (cx <- 0 until nCircles) yield outerRadius / (nCircles-cx)
+    val groupSizes = for (cx <- 0 until nCircles) yield (cx + 1) * nPoints
+    var ix = 0
+    val points = for (cx <- 0 until nCircles;
```

Comment:
Same for `points` about `yield` usage. This line could be rewritten as

~~~scala
def generateCircle(radius: Double, n: Int): Seq[(Double, Double)] = {
  val theta = 2.0 * math.Pi / n
  Seq.tabulate(n)(i => (radius * math.cos(i * theta), radius * math.sin(i * theta)))
}

val points = (0 until nCircles).flatMap { i =>
  generateCircle(radii(i), groupSizes(i))
}.zipWithIndex
~~~

which is easier to read.